bin/octocatalog-diff

```diff
@@ -47,46 +67,84 @@ def get_hosts_from_puppetdb():
-            # are reprovisioned, etc. and change frequently
-            '--ignore', "'Sshkey[*]'",
+        ' '.join(
+            (
+                'octocatalog-diff',
+                # Add a --debug flag to get much more verbose output
+                '-n', host,
+                '--enc-override',
+                'environment=production,parameters::dummy_secrets=true',
+                # Unfortunately color codes are not handled well by github, so
+                # we turn those off here
+                '--no-color',
+                '--display-detail-add',
+                # Ignore changes to puppetserver settings, since these are
+                # dependent on the path that octocatalog-diff sets and thus
+                # always differ
+                "--ignore 'Ini_setting[puppet.conf/master/storeconfigs]'",
+                "--ignore 'Ini_setting[puppet.conf/master/storeconfigs_backend]'",
+                "--ignore 'Ini_setting[puppetdbserver_urls]'",
+                "--ignore 'Ini_setting[soft_write_failure]'",
+                "--ignore 'File[/tmp/*/routes.yaml]'",
+                # Ignore changes to SSH keys, since these change when any
+                # machines are reprovisioned, etc. and change frequently
+                "--ignore 'Sshkey[*]'",
+            ),
         ),
         stderr=subprocess.PIPE,
         stdout=subprocess.PIPE,
+        # We need shell=True here for the diffs to calculate correctly on the
+        # puppetmaster, otherwise the ignore settings above do not take effect.
+        # This should be safe enough as we aren't accepting user-defined input
+        shell=True,
     )
 
+    stdout = process.stdout.decode('utf-8').rstrip()
+    stderr = process.stderr.decode('utf-8').rstrip()
+    info('stdout for {}'.format(host))
+    info(stdout)
+    info('stderr for {}'.format(host))
+    info(stderr)
+
     if process.returncode == 0:
         # This is a clean run, puppet did not error and there was no diff
-        print('run for {}: clean'.format(host))
-        return 0
+        return (host, 0, None)
     elif process.returncode == 2:
         # This means the puppet run has found some diff from master, so we
-        # should keep this and report it back in the review
-        print(process.stdout.decode('utf-8').rstrip())
-        return 0
+        # return this and report it back in the review but do not fail the run
+        return (host, 0, stdout)
     else:
         # This means the puppet run has failed in some way (a return code of 1
-        # is pretty usualfor that), so the whole command should error out
-        print('run for {} errored:'.format(host))
-        print(process.stderr.decode('utf-8').rstrip())
-        return 1
+        # is pretty usual for that), so we return an error code and the output
+        # to post in the review
+        return (host, 1, GITHUB_COLLAPSE_CODE.format('**error** for {}'.format(host), 'text', stderr))
```

Comment:
does it make sense to include the stdout here as well?

Reply:
I don't think the stdout has anything particularly useful when errors like this happen (it's blank for instance for all the errors on [build 15](https://jenkins.ocf.berkeley.edu/blue/rest/organizations/jenkins/pipelines/puppet/branches/PR-762/runs/15/nodes/42/steps/47/log/?start=0) where I added a test error), and it is printed shortly above this (line 108), so it'll still be in the jenkins log itself.